src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,594 @@
+                throw new NoSuchProviderException(msg);
+            }
+            return new KDF((KDFSpi) instance.impl, instance.provider, algorithm,
+                           kdfParameters);
+
+        } catch (NoSuchAlgorithmException nsae) {
+            return handleException(nsae);
+        }
+    }
+
+    /**
+     * Returns a {@code KDF} object that implements the specified algorithm from
+     * the specified provider and is initialized with the specified parameters.
+     *
+     * @param algorithm
+     *     the key derivation algorithm to use
+     * @param kdfParameters
+     *     the {@code KDFParameters} used to configure this KDF's algorithm or
+     *     {@code null} if no additional parameters are provided
+     * @param provider
+     *     the provider to use for this key derivation
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if the specified provider does not support the specified KDF
+     *     algorithm
+     * @throws InvalidAlgorithmParameterException
+     *     if the {@code KDFParameters} is an invalid value
+     * @throws NullPointerException
+     *     if the algorithm or provider is {@code null}
+     */
+    public static KDF getInstance(String algorithm,
+                                  KDFParameters kdfParameters,
+                                  Provider provider)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Objects.requireNonNull(algorithm, "null algorithm name");
+        Objects.requireNonNull(provider, "provider may not be null");
+        try {
+            Instance instance = GetInstance.getInstance("KDF", KDFSpi.class,
+                                                        algorithm,
+                                                        kdfParameters,
+                                                        provider);
+            if (!JceSecurity.canUseProvider(instance.provider)) {
+                String msg = "JCE cannot authenticate the provider "
+                             + instance.provider.getName();
+                throw new SecurityException(msg);
+            }
+            return new KDF((KDFSpi) instance.impl, instance.provider, algorithm,
+                           kdfParameters);
+
+        } catch (NoSuchAlgorithmException nsae) {
+            return handleException(nsae);
+        }
+    }
+
+    private static KDF handleException(NoSuchAlgorithmException e)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Throwable cause = e.getCause();
+        if (cause instanceof InvalidAlgorithmParameterException) {
+            throw (InvalidAlgorithmParameterException) cause;
+        }
+        throw e;
+    }
+
+    /**
+     * Derives a key, returned as a {@code SecretKey}.
+     * <p>
+     * The {@code deriveKey} method may be called multiple times on a particular
+     * {@code KDF} instance.
```

Comment:
Similar comment as in `KDFSpi`. If we don't want to talk about thread-safety, I'd rather not mention anything.